### PR TITLE
feat: sharded page cache + LRU

### PR DESCRIPTION
**Pull request by rphmeier**
This shards the page cache based on the number of workers, and gives each shard an LRU cache.

A further step would be to give each worker a mutable handle to their shard during `update`.

not implemented: cache eviction during retrieval.

### CODE REVIEW

**Comment by rphmeier** [comment]
* **#191** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/191?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/page_cache.rs [hunk 1]
```diff
@@ -300,39 +306,121 @@ impl PageStore {
...
+                cached: LruCache::unbounded_with_hasher(FxBuildHasher::default()),
+            }),
+            // UNWRAP: both factors are non-zero
+            page_limit: NonZeroUsize::new(PAGE_LIMIT_PER_ROOT_CHILD * count).unwrap(),
+        })
+        .collect()
+}
+
+/// The index of a shard of a page cache.
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Comment (pepyakin) [inline nomt/src/page_cache.rs:384]: Copy?

File: nomt/src/page_cache.rs [hunk 2]
```diff
@@ -343,25 +431,53 @@ impl PageCache {
...
                 fetch_tp,
             }),
         }
     }
 
+    fn shard_index_for(&self, page_id: &PageId) -> Option<usize> {
+        if page_id == &ROOT_PAGE_ID {
+            None
+        } else {
+            // note: this is ALWAYS `Some` because the whole page space is covered.
```
Comment (pepyakin) [inline nomt/src/page_cache.rs:452]: fail fast!

File: nomt/src/page_cache.rs [hunk 3]
```diff
@@ -449,56 +560,63 @@ impl PageCache {
...
+
+        let mut shard = self.shard(shard_index).locked.lock();
+        if let Some(page) = shard.cached.get(&page_id) {
+            return Page {
+                inner: page.clone(),
+            };
+        }
+
+        let shard_mut = &mut *shard;
+        let existing_inflight = if let Some(inflight) = shard_mut.inflight.get(&page_id) {
```
Comment (pepyakin) [inline nomt/src/page_cache.rs:579]: I think `existing_inflight` thing is not really elegant and a better solution would be to explicitly call `drop(shard_guard)` and then call wait within the if statement. You would have to clone out the inflight beforehand.

**Review by pepyakin — APPROVED** [review]

File: nomt/src/page_cache.rs [hunk 4]
```diff
@@ -300,39 +306,121 @@ impl PageStore {
...
+// Each shard has its own domain and handles a sub-tree of the page tree, defined by a
+// continuous set of children of the root page.
+struct CacheShard {
+    region: PageRegion,
+    locked: Mutex<CacheShardLocked>,
+    page_limit: NonZeroUsize,
+}
+
+struct CacheShardLocked {
+    inflight: FxHashMap<PageId, Arc<InflightFetch>>,
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply nomt/src/page_cache.rs:287]: ah, I didn't realize

File: nomt/src/page_cache.rs (hunk 1 above)
Comment (rphmeier) [reply nomt/src/page_cache.rs:353]: okay

File: nomt/src/page_cache.rs (hunk 2 above)
Comment (rphmeier) [reply nomt/src/page_cache.rs:421]: ok

File: nomt/src/page_cache.rs (hunk 3 above)
Comment (rphmeier) [reply nomt/src/page_cache.rs:548]: can we punt on this? it was left over from before and we may refactor I/O more later

Comment (pepyakin) [reply nomt/src/page_cache.rs:552]: ofc!

**Pull request merged**